feat(cloudtrail): add configuration for IsOrganizationTrail

daschaa opened the pull request:
Fixes #21578

Please add `pr-linter/exempt-readme` label since this property needs no entry in the README imho.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 768f21189009006c99120c2c5130a5928bb2cc66
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9b603b57-a76d-4c1d-8853-11d1cb12ff5d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: a4819f87db2af893ab55265c6f18c2abfc429c69
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3efb7b16-21f9-4acd-bd7d-1e2e2899602f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra commented on packages/@aws-cdk/aws-cloudtrail/lib/cloudtrail.ts:
```diff
@@ -115,6 +115,17 @@ export interface TrailProps {
    */
   readonly bucket?: s3.IBucket;
+
+  /**
+   * Specifies whether the trail is applied to all accounts in an organization in AWS Organizations, or only for the current AWS account.
+   * The default is false, and cannot be true unless the call is made on behalf of an AWS account that is the management accountfor an organization in AWS Organizations.
+   * If the trail is not an organization trail and this is set to true, the trail will be created in all AWS accounts that belong to the organization.
+   * If the trail is an organization trail and this is set to false, the trail will remain in the current AWS account but be deleted from all member accounts in the organization.
+   *
```
This documentation feels a little clunky to me. After the first read through I felt like I didn't quite get it so I went and looked up `CfnTrail` to try to get clarification. I see that it came from there but I think we can make this more clear.

TheRealAmazonKendra commented on packages/@aws-cdk/aws-cloudtrail/lib/cloudtrail.ts:
```diff
@@ -285,6 +296,7 @@ export class Trail extends Resource {
       cloudWatchLogsRoleArn: logsRole?.roleArn,
       snsTopicName: this.topic?.topicName,
       eventSelectors: this.eventSelectors,
+      isOrganizationTrail: props.isOrganizationTrail ?? false,
```
What's the behavior here if the account isn't the owner of an organization and this gets set to true? Does it just fail silently? If so, do we have a way to validate that the account is actually the owner of an organization?

TheRealAmazonKendra replied:
```suggestion
      isOrganizationTrail: props.isOrganizationTrail,
```
Let's just pass in what's provided. That way it doesn't impact templates unless it's explicitly provided.

TheRealAmazonKendra reviewed (changes requested):
I think I would actually like a README update on this because the CloudFormation documentation isn't clear enough. I would have been inclined to agree with you if it hadn't taken me multiple times reading it to understand what the actual behavior was.

daschaa commented:
@TheRealAmazonKendra I have added an entry in the README, updated the docstring and implemented your suggestion to not use the `?? false` operator. Thanks for your valuable feedback!
Regarding the check if an account manages an organization, i did not find anything that could solve this problem. Do you know any way to check this?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 6f324d06dcc784f966bd3a92fcedc1e2eac3f20a
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7ed0781f-c346-4344-9c6d-0261ed4b411a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 23c483cacff98302b1d186965dcc2abd5386eba3
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=74405aff-4309-41c6-a49a-12d82d8bed1e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra commented:
> @TheRealAmazonKendra I have added an entry in the README, updated the docstring and implemented your suggestion to not use the `?? false` operator. Thanks for your valuable feedback! Regarding the check if an account manages an organization, i did not find anything that could solve this problem. Do you know any way to check this?

I'm not sure that there is a good way without introducing a LOT of extra complexity to this.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 3b8ffe078b357e860a6ac486a44b605046490c02
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0dc91488-efeb-45fd-a756-383405b6dc8d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
